File: algorithms/0033/main.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
  const rotateIdx = findRotateIdx(nums)
  const len = nums.length
  let left = 0
  let right = len - 1
  while (left <= right) {
    const mid = Math.floor((left + right) / 2)
    const rotateMid = (rotateIdx + mid) % len
    if (nums[rotateMid] < target) {
      left = mid + 1
    } else if (target < nums[rotateMid]) {
      right = mid - 1
    } else {
      return rotateMid
    }
  }
  return -1
}
function findRotateIdx(nums) {
  let l = 0
  let r = nums.length - 1
  while (l < r) {
    const mid = Math.floor((l + r) / 2)
    nums[r] < nums[mid] ? (l = mid + 1) : (r = mid)
  }
  return l
}
module.exports = search

File: algorithms/0033/main.go
package p0033

func search(nums []int, target int) int {
	rotateIdx := rotateIdx(nums)
	length := len(nums)
	left, right := 0, length-1
	for left <= right {
		mid := (left + right) / 2
		/* nums 是 rotated，所以需要使用 rotatedMid 来获取 mid 的值 */
		rotatedMid := (rotateIdx + mid) % length
		switch {
		case nums[rotatedMid] < target:
			left = mid + 1
		case target < nums[rotatedMid]:
			right = mid - 1
		default:
			return rotatedMid
		}
	}
	return -1
}

// 找到开始旋转的那个点的索引
func rotateIdx(nums []int) int {
	l, r := 0, len(nums)-1
	for l < r {
		mid := (l + r) / 2
		if nums[r] < nums[mid] {
			l = mid + 1
		} else {
			r = mid
		}
	}
	return l
}
